Unstable merge: asset library with transparent compression, plus a stream of small changes

rasky opened the pull request:
This PR imports the asset library from the unstable branch. This is the subject of the first half of commits.

The asset library allows to load binary files from any filesystem with transparent compression, and is the basis of all data files management in the unstable branch, which is in fact compressed.

Compression works by wrapping a data file (think gzip), so `asset_load()` will transparently decompress it while loading it into RDRAM. Moreover, `asset_fopen()` uses some libc magic to create a `FILE*` that transparently decompresses while reading, to allow compressing data files that need to be parsed at runtime.

There are two compression levels at the moment: 

 * Level 1 is based on LZ4, which features incredible fast decompression speed with decent compression ratios. It is so fast that it's actually faster to load and decompress a compressed file, rather than loading it uncompressed (!). On the unstable branch, we compress all assets by default with level 1 because of this.
 * Level 2 is based on LZH5, part of the LHA family. It offers better-than-gzip ratios on small files, with a slower decompression. The decompressor is based on open source code that has been optimized for N64, but it's still not as fast as it could. Specifically, the Huffman decompression is still made bitwise rather than bytewise and could use some speed up.

To compress a binary file, the user can use the new tool `mkasset`. In the future, asset-generating tools (like the new version of mksprite) will transparently offer compression capabilities.

See this table for more data (c0 = uncompress, c1 = lz4, c2 = lzh5)
```
File                        : KiB  | Ratio   | Load+Dec (ms)
--------------------------------------------------------------
fc_200-fs8.sprite.c0        :   31 |  100.0% |     11.4 
fc_200-fs8.sprite.c1        :   31 |   32.9% |      5.9 
fc_200-fs8.sprite.c2        :   31 |   25.7% |     19.6 
font.sprite.c0              :   22 |  100.0% |      7.8 
font.sprite.c1              :   22 |   19.8% |      4.1 
font.sprite.c2              :   22 |   14.4% |      9.4 
hoi-fs8.sprite.c0           :  125 |  100.0% |     41.1 
hoi-fs8.sprite.c1           :  125 |   11.5% |     14.3 
hoi-fs8.sprite.c2           :  125 |    8.9% |     29.3 
lens_64-fs8.sprite.c0       :   16 |  100.0% |      7.1 
lens_64-fs8.sprite.c1       :   16 |   26.7% |      4.4 
lens_64-fs8.sprite.c2       :   16 |   23.8% |      9.4 
lenspic_256-fs8.sprite.c0   :   64 |  100.0% |     22.0 
lenspic_256-fs8.sprite.c1   :   64 |   50.5% |     10.5 
lenspic_256-fs8.sprite.c2   :   64 |   44.2% |     58.2 
prax4.fli.c0                :  327 |  100.0% |    104.5 
prax4.fli.c1                :  327 |   63.2% |     85.7 
prax4.fli.c2                :  327 |   51.2% |    316.9 
room_normal.ci8.sprite.c0   :  125 |  100.0% |     41.6 
room_normal.ci8.sprite.c1   :  125 |   31.7% |     19.2 
room_normal.ci8.sprite.c2   :  125 |   26.5% |     67.2 
srtitle400-fs8.sprite.c0    :  125 |  100.0% |     40.9 
srtitle400-fs8.sprite.c1    :  125 |   28.6% |     18.4 
srtitle400-fs8.sprite.c2    :  125 |   22.9% |     60.8 
troll-fs8.sprite.c0         :  125 |  100.0% |     41.9 
troll-fs8.sprite.c1         :  125 |   32.6% |     25.1 
troll-fs8.sprite.c2         :  125 |   24.7% |     61.5 
```

The rest of the PR is a stream of small semi-obvious commits that just minimizes the differences with unstable.

rasky committed:
dma: add PI registers definitions

rasky committed:
Add new asset library

This library handles loading asset files from filesystems. It allows
either one-time loads or incremental loads via FILE* interface, and
it handles transparent decompression in both cases.

It implements two compressed format: a very fast one (based on LZ4),
and a slower one (based on LZH5) that compresses files more.

rasky committed:
asset.h: improve documentation

rasky committed:
tools: print tool name during compilation

rasky committed:
tools: add mkasset tool to compress assets

rasky committed:
tools: add asset compression library

This library includes common code for asset compression. It relies
on the LZH5 compression library (previously moved from the audio
library), plus a vendored copy of the LZ4 compressor.

rasky committed:
tools: add simple library to write big-endian binary files.

This will be use to simplify all tools writing binary files that are
now byteswapping data in memory (and often causing weird bugs because
of that).

rasky committed:
lzh5_compress: fix MinGW compilation by adding missing include

rasky committed:
debugcpp.h: fix pasto

rasky committed:
cop1.h: add C1_ENABLE_MASK / C1_CAUSE_MASK to group bits

rasky committed:
dragonfs: improve management of error codes

This commit does three things that are related:

 * Rename DFS_ENOMEM to DFS_ENFILE to more closely match the equivalent
   POSIX error (ENFILE)
 * Add dfs_strerror() to simplify reporting errors produced by dfs
 * Correctly set errno on fopen() failure, so that callers can find out
   why a fopen() failed

rasky committed:
regsinternal.h: add missing include

rasky committed:
entrypoint.S: simplify readability of initial SP constant

rasky committed:
interrupt: move static variable to local scope

rasky committed:
inthandler: fix reference to reg_block_t in comment

rasky committed:
exception: avoid using __builint_isnan as well

It does seem to misbehave depending on compilation flags so it seems
better to manually check for bit patterns rather than playing further
cat and mouse games with the compiler.

rasky committed:
n64.mk: fix quoting of currnet directory in command line

rasky committed:
Improve compilation with NDEBUG

Currently, a NDEBUG build is broken because of compilation and linking
errors. It is currently not tested by CI. This commit fixes a few issues
and make sure it works. Going forward, we should probably cover with
CI if we want to make sure it always works.

rasky committed:
xm64/ym64: use must_fopen

rasky committed:
system.c: improve errno usage in open()

Currently open() was a bit sloppy in errno usage and specifically
it didn't allow the filesystem open call to set it. Change it so
that filesystem implementations can now set it to propagate a detailed
error message to the caller.

Also remove some weird global errno redefinition, not sure why it
was ever added but everything seems fine without it.

rasky committed:
testrom: avoid redefining rand() symbol

snacchus commented on include/asset.h:
```diff
@@ -0,0 +1,106 @@
+ * 
+ * ## Asset compression
+ * 
+ * All tools in libdragon that produce assets already integrate compression
+ * support compatible with the asset libraries. To compress your own data
```
Is this actually true on trunk right now?

rasky replied:
removed that sentence

snacchus commented on include/asset.h:
```diff
@@ -0,0 +1,106 @@
+ * automatically uncompressed.
+ * 
+ * @param fn        Filename to load (including filesystem prefix)
+ * @param sz        Pointer to an integer where the size of the file will be stored
```
`sz` can be `NULL`, right? Should probably mention this here.

rasky replied:
Done

snacchus commented on src/compress/ringbuf.c:
```diff
@@ -0,0 +1,64 @@
+    int ringbuf_copy_pos = (ringbuf->ringbuf_pos - copy_offset) & (RING_BUFFER_SIZE-1);
+    int dst_pos = 0;
+    while (count > 0) {
+		int wn = count;
+        wn = wn < RING_BUFFER_SIZE - ringbuf_copy_pos     ? wn : RING_BUFFER_SIZE - ringbuf_copy_pos;
+		wn = wn < RING_BUFFER_SIZE - ringbuf->ringbuf_pos ? wn : RING_BUFFER_SIZE - ringbuf->ringbuf_pos;
```
Some whitespace issues here

rasky replied:
Fixed

meeq committed:
Fix Make 3.8 compatibility issues in examples Makefile

rasky committed:
compress: fix tabs/spaces mixup

rasky committed:
Add new asset library

This library handles loading asset files from filesystems. It allows
either one-time loads or incremental loads via FILE* interface, and
it handles transparent decompression in both cases.

It implements two compressed format: a very fast one (based on LZ4),
and a slower one (based on LZH5) that compresses files more.

rasky committed:
lzh5: remove old LZH5 implementation (local to the audio library)

rasky commented:
Added also the new API `asset_init_compression()` as discussed on Discord, to avoid linking all compression levels unless really necessary.
